src/us/kbase/workspace/database/provenance/CreditMetadata.java

```diff
@@ -0,0 +1,345 @@
+				resourceType,
+				titles,
+				version);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		CreditMetadata other = (CreditMetadata) obj;
+		return Objects.equals(comments, other.comments)
+				&& Objects.equals(contributors, other.contributors)
+				&& Objects.equals(dates, other.dates)
+				&& Objects.equals(funding, other.funding)
+				&& Objects.equals(identifier, other.identifier)
+				&& Objects.equals(license, other.license)
+				&& Objects.equals(relatedIdentifiers, other.relatedIdentifiers)
+				&& Objects.equals(resourceType, other.resourceType)
+				&& Objects.equals(titles, other.titles)
+				&& Objects.equals(version, other.version);
+	}
+
+	/**
+	 * Gets a builder for the {@link CreditMetadata}.
+	 *
+	 * @param identifier unique persistent ID for the resource
+	 * @param resourceType type of the resource, as a {@link ResourceType}
+	 * @param contributors list of {@link Contributor} objects
+	 * @param titles list of {@link Title} objects
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(
+			final String identifier,
+			final ResourceType resourceType,
+			final List<Contributor> contributors,
+			final List<Title> titles) {
+		return new Builder(identifier, resourceType, contributors, titles, new ArrayList<>());
+	}
+
+	/**
+	 * Gets a builder for the {@link CreditMetadata}.
+	 *
+	 * @param identifier unique persistent ID for the resource
+	 * @param resourceType type of the resource as a string
+	 * @param contributors list of {@link Contributor} objects
+	 * @param titles list of {@link Title} objects
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(
+			final String identifier,
+			final String resourceType,
+			final List<Contributor> contributors,
+			final List<Title> titles) {
+		final List<String> errorList = new ArrayList<>();
+
+		ResourceType rt = null;
+		if (resourceType != null) {
+			try {
+				rt = ResourceType.getResourceType(resourceType);
+			} catch (IllegalArgumentException e) {
+				// TEMPORARY HACK
+				// As there is only one resource type at present (ResourceType.DATASET), null input
+				// in the resourceType field will be automatically set to that value.
+				// The error message "resourceType cannot be null or whitespace only" thus
+				// needs to be edited to say "resourceType cannot be whitespace only".
+				String errorMessage = e.getMessage();
+				if ("resourceType cannot be null or whitespace only".equals(errorMessage)) {
+					errorMessage = "resourceType cannot be whitespace only";
+				}
+				errorList.add(errorMessage);
+			}
+		}
+		return new Builder(identifier, rt, contributors, titles, errorList);
+	}
+
+	/** A builder for {@link CreditMetadata}. */
+	public static class Builder {
+		private String comments;
+		private String identifier;
```

Comment:
Mark required fields as final

Reply:
`identifier` should be required right?

Reply:
Yes, but it needs to be massaged with `checkPid` first ==> it isn't set as `final`

Reply:
I don't think I understand - you can do whatever you want to the data you're assigning to a final instance variable before you assign it, you just can't assign it more than once. Looking at the builder constructor it looks to me like making the instance variable `final` would be fine

Reply:
:+1: